File: lessions/lession-10/sort-an-array.py
# class Solution:
#     def sortArray(self, nums: List[int]) -> List[int]:
#       self.mergeSort(nums, 0, len(nums) - 1)
#       return nums

#     def mergeSort(self, nums: List[int], l: int, r: int):
#       if l >= r:
#         return
#       mid = (l + r) >> 1
#       self.mergeSort(nums, l, mid)
#       self.mergeSort(nums, mid + 1, r)
#       self.merge(nums, l, mid, r)

#     def merge(self, nums: List[int], l: int, mid: int, r: int):
#       # left [l, mid]
#       # right [mid + 1, r]
#       length = r - l + 1
#       temp = []
#       i = l
#       j = mid + 1
#       for k in range(length):
#         if (j > r or (i <= mid and nums[i] < nums[j])):
#           temp.append(nums[i])
#           i += 1
#         else:
#           temp.append(nums[j])
#           j += 1

#       nums[l:r+1] = temp



def mergesort(nums, left, right):
  if right <= left:
    return
  mid = (left+right) >> 1
  mergesort(nums, left, mid)
  mergesort(nums, mid+1, right)
  merge(nums, left, mid, right)


def merge(nums, left, mid, right):
  temp = []
  i = left
  j = mid+1
  while i <= mid and j <= right:
    if nums[i] <= nums[j]:
        temp.append(nums[i])
        i +=1
    else:
      temp.append(nums[j])
      j +=1
  while i<=mid:
    temp.append(nums[i])
    i +=1
  while j<=right:
    temp.append(nums[j])
    j +=1
  nums[left:right+1] = temp

File: lessions/lession-10/sort-an-array.ts
function partition(arr: number[], l: number, r: number): number {
  let pivot = l + Math.trunc(Math.random() * (r - l + 1));
  let pivotVal = arr[pivot];
  while (l <= r) {
    while (arr[l] < pivotVal) l++;
    while (arr[r] > pivotVal) r--;
    if (l === r) break;
    if (l < r) {
      let temp = arr[l];
      arr[l] = arr[r];
      arr[r] = temp;
      l++;
      r--;
    }
  }
  return r;
}

function quickSort(arr: number[], l: number, r: number) {
  if (l >= r) return;
  let pivot = partition(arr, l, r);
  quickSort(arr, l, pivot);
  quickSort(arr, pivot + 1, r);
}

function sortArray(nums: number[]): number[] {
  quickSort(nums, 0, nums.length - 1);
  return nums;
}
